ostd/src/mm/frame/linked_list.rs

```diff
@@ -0,0 +1,674 @@
+        debug_assert!(frame.meta_mut().next_mut().is_none());
+        debug_assert!(frame.meta_mut().prev_mut().is_none());
+        debug_assert_eq!(frame.slot().in_list.load(Ordering::Relaxed), 0);
+
+        let frame_ptr = NonNull::from(frame.meta_mut());
+
+        if let Some(current) = &mut self.current {
+            // SAFETY: We have ownership of the links via `&mut self`.
+            unsafe {
+                if let Some(prev) = current.as_mut().prev.get_mut() {
+                    debug_assert_eq!(*prev.as_mut().next.get_mut(), Some(*current));
+                    *prev.as_mut().next.get_mut() = Some(frame_ptr);
+                    *frame.meta_mut().prev.get_mut() = Some(*prev);
+                    *frame.meta_mut().next_mut() = Some(*current);
+                    *prev = frame_ptr;
+                } else {
+                    debug_assert_eq!(self.list.front, Some(*current));
+                    *frame.meta_mut().next_mut() = Some(*current);
+                    *current.as_mut().prev.get_mut() = Some(frame_ptr);
+                    self.list.front = Some(frame_ptr);
+                }
+            }
+        } else {
+            // We are at the "ghost" non-element.
+            if let Some(back) = &mut self.list.back {
+                // SAFETY: We have ownership of the links via `&mut self`.
+                unsafe {
+                    debug_assert!(back.as_mut().next.get_mut().is_none());
+                    *back.as_mut().next.get_mut() = Some(frame_ptr);
+                }
+                *frame.meta_mut().prev.get_mut() = Some(*back);
+                self.list.back = Some(frame_ptr);
+            } else {
+                debug_assert_eq!(self.list.front, None);
+                self.list.front = Some(frame_ptr);
+                self.list.back = Some(frame_ptr);
+            }
+        }
+
+        frame
+            .slot()
+            .in_list
+            .store(self.list.lazy_get_id(), Ordering::Relaxed);
+
+        // Forget the frame to transfer the ownership to the list.
+        let _ = frame.into_raw();
+
+        self.list.size += 1;
+    }
+
+    /// Provides a reference to the linked list.
+    pub fn as_list(&self) -> &LinkedList<M> {
+        &*self.list
+    }
+}
+
+impl<M> Drop for LinkedList<M>
+where
+    Link<M>: AnyFrameMeta,
+{
+    fn drop(&mut self) {
+        let mut cursor = self.cursor_front_mut();
+        while cursor.take_current().is_some() {}
+    }
+}
+
+/// The metadata of linked list frames.
+///
+/// To allow other metadata to be customized, this type is a wrapper around the
+/// actual metadata type `M`.
+///
+/// Linked list frames can be contained in a [`LinkedList`].
+#[derive(Debug)]
+pub struct Link<M> {
+    // The prev and next pointers are mutably borrowed when the list is mutably
+    // borrowed.
+    next: UnsafeCell<Option<NonNull<Link<M>>>>,
+    prev: UnsafeCell<Option<NonNull<Link<M>>>>,
```

Comment:
It seems you've changed the linked list implementation to include `UniqueFrame` (instead of `Frame` in the original implementation), right? So during insertion and removal you have a mutable reference to `Link<M>`. That means you don't really need internal mutability now.

I think we can remove `UnsafeCell` here and just do it:
```patch
-                     *prev.as_mut().next.get_mut() = Some(frame_ptr);
+                     prev.as_mut().next = Some(frame_ptr);
```